### ISSUE TITLE
Add `nesting-selector-no-missing-scoping-root`

### ISSUE THREAD

**Issue by ekilah** [stylelint/stylelint#8406]
### What is the problem you're trying to solve?

I want to prevent the use of unscoped nesting with `&`.

For example, we have a lint rule (`selector-max-type`) that prevents top-level styles on HTML tags like `li` or `table`:

```js
{
    'selector-max-type': [
      0,
      {
        ignore: ['descendant', 'child', 'compounded', 'next-sibling'],
      },
    ],
}
```

That works most of the time, but a rule like `& li { }` at the top level was not caught because of the `&`. This allowed all `li`s to be styled in the application accidentally.

This was a request over in https://github.com/stylelint/stylelint/issues/2255#issuecomment-1220333960 but was not followed up on with a new issue.

I asked about it here in the proposal for blocking redundant nesting: https://github.com/stylelint/stylelint/issues/8162#issuecomment-2669132433, but @romainmenke helpfully pointed out that top-level nesting `&`s aren't actually _redundant_. That said, I still want a rule that can prevent them.

### What solution would you like to see?

I would like to prevent nesting `&`s that are at the top level (as in, `&`s that are not contained inside another rule)

These should fail:

```css
& { }

& li { } 

& .foo { } 

&.foo { }

&:hover { }
```

These should pass:

```css
.foo {
  & .bar { }
}

.foo { 
  & { } 
}
```

The last example above may be redundant, but wouldn't be the job of this rule to catch. That'd be left to

- https://github.com/stylelint/stylelint/issues/8162

**Comment by Mouvedia** [comment]
If we could find more cases, it may be worth adding.

e.g. `no-implicit-root`
```css
// invalid at depth 0

& a { }
:scope a {}
@scope (html) {}
@scope (svg[xmlns]) {}

// valid

:root a {}
a {}
:host a {} /* e.g. https://v17.angular.io/guide/component-styles#host */
@scope (.foo) {
  :scope a {}
}
```

**Comment by ekilah** [comment]
> I think the motivation doesn't quite fit the criteria for a core rule even when I think we can create a rule that warns for the same cases.
> In other words we should not discourage most developers from writing top-level HTML tags.

To be clear, the motivation I gave in the OP re: `selector-max-type` was just provided as an example where top-level/unscoped `&`s are causing _me_ problems. It's definitely not the _only_ reason these unscoped `&`s could cause problems, or the only use case where this rule would be useful. 

The intent was perhaps more like "unscoped `&`s seem unideal and/or likely indicate a mistake" but I went further with something like "and their presence can end up breaking other rules' detection mechanisms, like this example."

Sorry if my initial example was confusing/misleading. The rule itself doesn't have to know about my `selector-max-type` problem, but it'd help fix it.

I agree that not that everyone should ban top-level HTML tag styling, though I think there are definitely good use cases for wanting to (e.g. if you are using CSS modules, like I am). That's mostly beside the point, though, and I can concede that this may fit under "idiosyncratic".

------

> Maybe a block was copy/pasted and the developer left the `&` in place?

Agreed! This is a very easy mistake to make.


> This also happens to cover the cases as laid out by @ekilah even when the intention is a bit different.
> This should also fail (`@media { & {} }, etc)

Agreed, there are probably more cases like that that are also unlikely to be intentional!

I didn't mention all the various `@` rules in my example ideal failures because I don't know the nuances of all of them well enough to make suggestions about them, like your suggestion that `@scope (.foo) { & {} }` should still be allowed.

**Comment by Mouvedia** [comment]
> But specifically warning for a top-level or an unscoped conditional `&` might be a good core Stylelint rule.

What would be its name?

**Comment by romainmenke** [comment]
It is a **nested selector** that represents `:scope` while at the same time lacking a **scoping root**.

`nested-selector-no-missing-scoping-root`

**Comment by jeddy3** [comment]
Rule details:

- Name: `nesting-selector-no-missing-scoping-root`
- Description: Disallow missing scoping root for nesting selectors.
- Message: Unexpected missing scoping root
- Primary option: `true`
- Secondary options: None
- Category: Avoid errors > Missing
- Autofix?: No
- Included in sharable configs?: Yes, in ~recommended~standard

@ekilah I've labelled the issue as ready to implement. Please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to add a new rule](https://stylelint.io/developer-guide/rules#add-a-rule) in the Developer guide.

**Comment by sw1tch3roo** [comment]
I'll take this if you don't mind ( ͡° ͜ʖ ͡°)

**Comment by jeddy3** [comment]
That would be fantastic, thank you!

**Comment by Mouvedia** [comment]
> Included in sharable configs?: Yes, in recommended

Are you sure ? I think standard would be more appropriate.

**Comment by sw1tch3roo** [comment]
I think standard is more appropriate too. The rule enforces modern CSS nesting conventions rather than preventing basic syntax errors, which aligns with the standard config's purpose to "enforce modern conventions found in the CSS specifications."

Since `& { }` at top-level is valid CSS but poor practice, it fits the standard config's focus on best practices rather than the recommended config's focus on error prevention.

**Comment by fimio-alex** [comment]
This triggers for  uses of [Tailwind's custom `@utility` directive](https://tailwindcss.com/docs/adding-custom-styles#complex-utilities).

**Comment by jeddy3** [comment]
@fimio-alex Tracked in https://github.com/stylelint/stylelint/issues/8728. That issue is marked as ready to implement. Please consider [contributing](https://stylelint.io/contributing) if you have time. There are [steps on how to add a new option](https://stylelint.io/developer-guide/rules#add-an-option-to-a-rule) in the Developer guide.
